osclib/stagingapi.py

```diff
@@ -1088,27 +1088,43 @@ def map_ring_package_to_subject(self, project, pkg):
 
         return project
 
-    def get_sub_packages(self, package, project=None):
+    def get_sub_packages(self, package, project):
         """
-        Returns a list of packages that need to be linked into rings
-        too. A package is actually a tuple of project and package name
+        Returns a list of packages that need to be linked into main package.
+        For adi package, check specfiles according to the main package.
         """
         ret = []
-        if not project:
+
+        # Started the logic. Note that, return empty tuple in case selecting
+        # non-ring package to a letter staging.
+        if self.is_adi_project(project):
+            # The additional check ensure the package is exist
+            if not self.item_exists(project, package):
+                return ret
+            # For adi package, do not trust the layout in the devel project, we
+            # must to guarantee the sub-pacakges are created according to the
+            # specfiles of main package. Therefore, main package must be
+            # created before through get_sub_packages().
```

Comment:
Is this check necessary as ring packages follow the same workflow? In fact on delete I imagine this would prevent proper cleanup.


```python
         delete_package(self.apiurl, subprj, package, force=True, msg=msg)
  
 -        for sub_prj, sub_pkg in self.get_sub_packages(package):
 +        for sub_prj, sub_pkg in self.get_sub_packages(package, project):
```

Meaning the package in `item_exists()` would be deleted just before `get_sub_packages()` is called. Since the source project is all that is needed and possible to be destroyed, perhaps just for deletes that should be checked. On submits it should create the project first (same as for rings) before linking to it. If the creation fails then the rest of the comments should probably crap just as they do for rings.

Reply:
Still not sure this check is good as it might just hide failure to properly stage a package, but re-ordering delete resolved that part.